### PR TITLE
Handle InstanceOfflinePing and close related clients

### LINKED ISSUE #416: SDK cannot detect some clients dropped when ASRS instance crashed.
Condition:
* ASRS instance crash.
* The ASRS instance for client connection is different from the instance for server connection.

Problem:
1. These clients will send ping, and app server do not notify offline for these clients.
2. No error will be throw.
3. These clients can be removed when server connection stopped.

### PR DESCRIPTION
**Pull request by JialinXin**
1. Store client-instance mapping from OpenConnectionMessages
2. Handle offline ping message to disconnect clients on crashes service instances

Fix #416, #539

### CODE REVIEW

File: src/Microsoft.Azure.SignalR.Common/ServiceConnections/ServiceConnectionBase.cs [hunk 1]
```diff
@@ -211,6 +213,10 @@ protected Task OnServiceErrorAsync(ServiceErrorMessage serviceErrorMessage)
 
         protected Task OnPingMessageAsync(PingMessage pingMessage)
         {
+            if (pingMessage.TryGetValue(Constants.ServicePingMessageKey.OfflineKey, out var instanceId) && !string.IsNullOrEmpty(instanceId))
+            {
+                return DisconnectClientConnections(instanceId);
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/ServiceConnections/ServiceConnectionBase.cs]: >DisconnectClientConnections [](start = 23, length = 27)

add some log

File: src/Microsoft.Azure.SignalR.AspNet/ServerConnections/ServiceConnection.cs [hunk 2]
```diff
@@ -278,14 +279,48 @@ private static string GetString(ReadOnlySequence<byte> buffer)
 
             return Encoding.UTF8.GetString(buffer.ToArray());
         }
-        
+
+        protected override Task DisconnectClientConnections(string instanceId)
+        {
+            if (_clientConnections.Count() == 0 || string.IsNullOrEmpty(instanceId))
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.AspNet/ServerConnections/ServiceConnection.cs]: >() [](start = 40, length = 2)

don't use the linq version `Count()`, use the property `Count`

File: src/Microsoft.Azure.SignalR.AspNet/ServerConnections/ServiceConnection.cs [hunk 3]
```diff
@@ -278,14 +279,48 @@ private static string GetString(ReadOnlySequence<byte> buffer)
...
+            {
+                return Task.CompletedTask;
+            }
+            try
+            {
+                // Connected service instance is down, close client gracefully
+                var connections = _clientConnections.Where(c => c.Value.InstanceId == instanceId);
+                foreach (var connection in connections)
+                {
+                    var serviceMessage = new CloseConnectionMessage(connection.Value.ConnectionId, errorMessage: "Service transient error");
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.AspNet/ServerConnections/ServiceConnection.cs]: >Service transient error [](start = 114, length = 23)

`The client is disconnected due to service internal error`?

File: src/Microsoft.Azure.SignalR.AspNet/ServerConnections/ServiceConnection.cs [hunk 4]
```diff
@@ -278,14 +279,48 @@ private static string GetString(ReadOnlySequence<byte> buffer)
...
+            catch (Exception ex)
+            {
+                Log.FailedToCleanupConnections(Logger, ex);
+            }
+            return Task.CompletedTask;
+        }
+
+        private string GetInstanceId(IDictionary<string, StringValues> header)
+        {
+            if (header.ContainsKey(Constants.AsrsInstanceId))
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.AspNet/ServerConnections/ServiceConnection.cs]: >ContainsKey [](start = 23, length = 11)

use TryGetValue to access the dict once

File: src/Microsoft.Azure.SignalR/ServerConnections/ServiceConnection.cs [hunk 5]
```diff
@@ -82,6 +83,30 @@ protected override async Task CleanupConnections()
...
+            }
+            try
+            {
+                // Connected service instance is down, close client gracefully
+                var connections = _connectionIds.Where(c => c.Value == instanceId);
+                foreach (var connection in connections)
+                {
+                    var serviceMessage = new CloseConnectionMessage(connection.Key, errorMessage: "Service transient error.");
+                    _ = PerformDisconnectAsyncCore(connection.Key, false);
+                    _ = WriteAsync(serviceMessage);
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR/ServerConnections/ServiceConnection.cs]: >WriteAsync [](start = 24, length = 10)

this is write back to the service so no need?

Comment (JialinXin) [reply src/Microsoft.Azure.SignalR/ServerConnections/ServiceConnection.cs]: I think it writes to clients, isn't it?

---
In reply to: [306695578](https://github.com/Azure/azure-signalr/pull/612#discussion_r306695578) [](ancestors = 306695578)

File: src/Microsoft.Azure.SignalR/ServerConnections/ServiceConnection.cs [hunk 6]
```diff
@@ -66,15 +67,20 @@ protected override Task DisposeConnection()
...
                 if (_connectionIds.Count == 0)
                 {
                     return;
                 }
-                await Task.WhenAll(_connectionIds.Select(s => PerformDisconnectAsyncCore(s.Key, false)));
+                var connectionIds = _connectionIds.Select(s => s.Key);
+                if (!string.IsNullOrEmpty(instanceId))
+                {
+                    connectionIds = _connectionIds.Where(s => s.Value == instanceId).Select(s => s.Key);
+                }
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR/ServerConnections/ServiceConnection.cs]:
```
var connectionIds = _connectionIds;

if (!string.IsNullOrEmpty(instanceId)){
connectionIds = _connectionIds.Where(s=>s.Value==instanceId);
}

 await Task.WhenAll(connectionIds.Select(s => PerformDisconnectAsyncCore(s.Key, false)));
```

File: test/Microsoft.Azure.SignalR.Tests/ServiceConnectionFacts.cs [hunk 7]
```diff
@@ -451,6 +452,52 @@ await proxy.WriteMessageAsync(new PingMessage()
...
+
+            // Server received instance offline ping on instanceId1 and trigger cleanup related client1
+            await proxy.WriteMessageAsync(new PingMessage()
+            {
+                Messages = new[] { "offline", instanceId1 }
+            });
+
+            // Validate client1 is closed and client2 is still connected
+            var disconnectTask = proxy.WaitForConnectionCloseAsync(connectionId1);
+            await disconnectTask.OrTimeout();
```
Comment (vicancy) [inline test/Microsoft.Azure.SignalR.Tests/ServiceConnectionFacts.cs]: >disconnectTask [](start = 18, length = 14)

any test to validate that disconnect task is handled?

File: src/Microsoft.Azure.SignalR/ServerConnections/ServiceConnection.cs (hunk 6 above)
Comment (JialinXin) [reply src/Microsoft.Azure.SignalR/ServerConnections/ServiceConnection.cs]: Tried before, but `Where` filter will return `Generic Dictionary` and can not convert to `ConcurrentDictionary`.

File: test/Microsoft.Azure.SignalR.Tests/ServiceConnectionFacts.cs (hunk 7 above)
Comment (JialinXin) [reply test/Microsoft.Azure.SignalR.Tests/ServiceConnectionFacts.cs]: current validation not enough? what else shall we check?

Comment (vicancy) [reply test/Microsoft.Azure.SignalR.Tests/ServiceConnectionFacts.cs]: I don't think this one checks the hub's disconnect event.

---
In reply to: [307564663](https://github.com/Azure/azure-signalr/pull/612#discussion_r307564663) [](ancestors = 307564663)

**Pull request merged**